File: Data Structures and Algorithms/C++/Queue.cpp
/*
    Ayush Kumar (202103103510253)
    4th Sem Div: A
    Queue.cpp
*/

#include "Queue.hpp"
#include <iostream>

using namespace std;

Queue::Queue(int size) {
  this->queue = new int[size];
  this->size = size;
}

void Queue::display(int index) {
  if (index < 0 || index > this->rear) {
    cout << "Queue out of index" << endl;
    return;
  }

  cout << this->queue[index] << endl;
}

void Queue::display() {
  cout << "{ ";
  for (int i = this->front + 1; i <= this->rear; i++) {
    cout << this->queue[i] << " ";
  }
  cout << "}" << endl;
}

void Queue::insert(int value) {
  if (this->rear >= this->size - 1) {
    cout << "Queue Overflow" << endl;
    return;
  }

  this->queue[++this->rear] = value;
}

void Queue::remove() {
  if (this->rear < 0) {
    cout << "Queue underflow" << endl;
    return;
  }

  this->queue[++this->front] = 0;

  if (this->rear == this->front) {
    this->front = -1;
    this->rear = -1;
  }
}

File: Data Structures and Algorithms/Python/Queue.py
class Queue:

  def __init__(self, size):
    self.__queue = []
    self.__rear = 0
    self.size = size

  def display(self, element = None):
    if(element == None):
      print(self.__queue)
    else:
      if(element <= 0 or element > len(self.__queue)):
        print("Queue out of index")
        return
      else:
        print(self.__queue[element - 1])

  def insert(self, value):
    if(self.__rear >= self.size):
      print("Stack overflow")
      return

    self.__queue.append(value)
    self.__rear += 1
      

  def remove(self):
    if(len(self.__queue) <= 0):
      print("Stack underflow")
      return

    self.__queue.pop(0)
    if len(self.__queue) == 0:
      self.__rear = 0
    
      